pkgs/development/compilers/julia/build-julia-package.nix

```
@@ -0,0 +1,158 @@
+  # Julia Artifacts need to be treated specially.
+  isJuliaArtifact = isJuliaArtifact;
+
+  inherit src;
+
+  inherit dontPatch patches patchPhase;
+
+  dontConfigure = true;
+
+  enableParallelBuilding = enableParallelBuilding;
+
+  requiredJuliaPackages = requiredJuliaPackages';
+
+  nativeBuildInputs = nativeBuildInputs';
+
+  buildInputs = buildInputs ++ requiredJuliaPackages' ++ [ stdenv.cc.libc ];
+
+  propagatedBuildInputs = propagatedBuildInputs;
+
+  preUnpack = ''
+    mkdir ${attrs.pname}
+    cd ${attrs.pname}
+  '';
+
+  setSourceRoot = ''
+    sourceRoot=$PWD
+  '';
+
+  # Julia installs compiled cache in the first path in DEPOT_PATH.
+  # Packages are pre-complied when imported.
+  buildPhase = if isJuliaArtifact then ''
+    runHook preBuild
+
+    mkdir -p $out/share/julia/artifacts/${attrs.juliaPath}
+    cp -r * $out/share/julia/artifacts/${attrs.juliaPath}
+
+    runHook postBuild
+  '' else ''
+    runHook preBuild
+
+    mkdir -p $out/share/julia/packages/${attrs.pname}
+    cp -r * $out/share/julia/packages/${attrs.pname}
+
+    export JULIA_LOAD_PATH=$out/share/julia/packages:${juliaLoadPath}:$JULIA_LOAD_PATH
+    export JULIA_DEPOT_PATH=$out/share/julia:${julia}/share/julia:${juliaDepotPath}
+
+    mkdir -p $out/share/julia/artifacts
+    for path in ${lib.concatMapStringsSep " " (p: p + "/share/julia/artifacts/*") juliaArtifacts}; do
+        ln -s $path $out/share/julia/artifacts/$(basename $path);
+    done
+
+    julia -e 'import ${attrs.pname}'
+
+    runHook postBuild
+  '';
+
+  # Need to install before building
+  dontInstall = true;
+
+  # Patch interpreter of bundled binary files
+  postFixup = if isJuliaArtifact  then ''
+    if [[ -d "$out"/share/julia/artifacts/${attrs.juliaPath}/bin ]]; then
+      for fn in $(echo $out/share/julia/artifacts/${attrs.juliaPath}/bin/*); do
+        if [[ $(file $fn) =~ .*[[:space:]]+ELF[[:space:]]+.*[[:space:]]+dynamically[[:space:]]linked,[[:space:]]interpreter ]]; then
```

There's an `isELF` function, but still we need a way to know if it is dynamically linked.

---

Yes, I started with `isELF` but it doesn't work as it doesn't distinguish between dynamically and statically linked binaries.

---

The other invocations of patchelf in nixpkgs seem to ignore the return value for good, i.e. `patchelf ... || true`, which is way easier?

---

`patchelf` gives an error if you try to change the interpreter of a statically linked binary. Isn't better to make a more accurate selection of files than to ignore errors?

---

No harm is done apart from spamming the log: https://github.com/NixOS/nixpkgs/blob/9269a0e15385ce1a1b007b3c901499eb499549cc/pkgs/development/tools/misc/patchelf/setup-hook.sh#L18